python/lsst/cp/pipe/bfCoeffCalc.py

```diff
@@ -0,0 +1,1272 @@
+                except NameError:
+                    self.debug.display = False
+                    self.log.warn('Failed to setup/connect to display! Debug display has been disabled')
+
+        plt.interactive(False)  # stop windows popping up when plotting. When headless, use 'agg' backend too
+        self.config.validateIsrConfig(self.log)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the BfTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitPairs", help="The list of visit pairs to use, as a list of tuples "
+                            "enclosed in quotes e.g. \"(123,456),(789,987),(654,321)\""
+                            "NB: must be comma-separated-tuples with no spaces, enclosed in quotes!")
+        parser.add_id_argument("--id", datasetType="bfKernelNew", ContainerClass=BfDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    @pipeBase.timeMethod
+    def run(self, dataRef, visitPairs):
+        """Run the brighter-fatter measurement task.
+
+        For a dataRef (which is each ccd here), and given a list of visit pairs, calulate the
+        brighter-fatter kernel for the ccd.
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the CCD for the visits to be fit.
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+        """
+        xcorrs = {}  # dict of lists keyed by either amp or CCD depending on level
+        means = {}
+        kernels = {}
+
+        # setup necessary objects
+        ccdNum = dataRef.dataId[self.config.ccdKey]
+        if self.config.level == 'CCD':
+            xcorrs = {ccdNum: []}
+            means = {ccdNum: []}
+        elif self.config.level == 'AMP':
+            # NB: don't use dataRef.get('raw_detector') due to composites
+            detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+            ampInfoCat = detector.getAmpInfoCatalog()
+            ampNames = [amp.getName() for amp in ampInfoCat]
+            xcorrs = {key: [] for key in ampNames}
+            means = {key: [] for key in ampNames}
+
+        # calculate or retrieve the gains
+        if self.config.doCalcGains:
+            self.log.info('Beginning gain estimation for CCD %s'%ccdNum)
+            gains, nomGains = self.estimateGains(dataRef, visitPairs)
+            dataRef.put(gains, datasetType='bfGain')
+            self.log.info('Finished gain estimation for CCD %s'%ccdNum)
+        else:
+            gains = dataRef.get('bfGain')
+            if not gains:
+                self.log.fatal('Failed to retrieved gains for CCD %s'%ccdNum)
+                raise RuntimeError("Must either calculate or supply gains for %s"%ccdNum)
```

Comment:
I actually don't know why I was logging a FATAL as well as raising the RuntimeError tbh. I moved the message from the log into the RuntimeError so that it's accurate, and removed the logger message. I think that should cover this.

Reply:
I agree. No point logging and then raising.